src/proof/supernova.rs

```diff
@@ -198,104 +195,84 @@ impl<'a, F: CurveCycleEquipped, C: Coprocessor<F> + 'a> SuperNovaProver<'a, F, C
+                RecursiveSNARK::new(
+                    &pp.pp,
+                    step,
+                    step,
+                    &secondary_circuit,
+                    z0,
+                    &Self::z0_secondary(),
+                )
+                .expect("failed to construct initial recursive SNARK")
+            });
+            info!("prove_step {i}");
+            recursive_snark
+                .prove_step(&pp.pp, step, &secondary_circuit)
+                .unwrap();
+            *rs = Some(recursive_snark);
+        };
 
         recursive_snark_option = if lurk_config(None, None)
             .perf
             .parallelism
             .wit_gen_vs_folding
             .is_parallel()
         {
-            let cc = steps
-                .into_iter()
-                .map(|mf| (mf.program_counter() == 0, Mutex::new(mf)))
-                .collect::<Vec<_>>();
+            // the sending end of the channel will block if it is at capacity
+            let (step_sender, step_receiver) = mpsc::sync_channel(MAX_BUFFERED_FRAMES);
 
             std::thread::scope(|s| {
-                s.spawn(|| {
-                    // Skip the very first circuit's witness, so `prove_step` can begin immediately.
-                    // That circuit's witness will not be cached and will just be computed on-demand.
-
-                    // There are many MultiFrames with PC = 0, each with several inner frames and heavy internal
-                    // parallelism for witness generation. So we do it like on Nova's pipeline.
-                    cc.iter()
-                        .skip(1)
-                        .filter(|(is_zero_pc, _)| *is_zero_pc)
-                        .for_each(|(_, mf)| {
-                            mf.lock()
-                                .unwrap()
-                                .cache_witness(store)
-                                .expect("witness caching failed");
-                        });
-
-                    // There shouldn't be as many MultiFrames with PC != 0 and they only have one inner frame, each with
-                    // poor internal parallelism for witness generation, so we can generate their witnesses in parallel.
-                    // This is mimicking the behavior we had in the Nova pipeline before #941 so...
-                    // TODO: once we have robust benchmarking for NIVC, we should test whether merging this loop with
-                    // the non-parallel one above (and getting rid of the filters) is better
-                    cc.par_iter()
-                        .skip(1)
-                        .filter(|(is_zero_pc, _)| !*is_zero_pc)
-                        .for_each(|(_, mf)| {
-                            mf.lock()
-                                .unwrap()
-                                .cache_witness(store)
-                                .expect("witness caching failed");
-                        });
```

Comment:
> Those only have one inner frame and it suffers from lack parallelism, unlike frames from multiframes with PC = 0. > That's because they're from handwritten coprocessor circuits, which are usually big circuit blobs.
> This is why this code does par_iter on them.

The code prior to this PR only starts doing witness caching for frames w/nonzero pc when the witness caching for those with zero-pc is done. It's highly non-obvious to me that this complex processing leads to faster witness caching, echoing [prior comments](https://github.com/lurk-lab/lurk-rs/pull/1011#discussion_r1445405735)

> I wouldn't touch this until we have proper benchmarks for SuperNova folding.

What's the issue for this, describing the gap precisely?

Reply:
> The code prior to this PR only starts doing witness caching for frames w/nonzero pc when the witness caching for those with zero-pc is done

Right, that's mostly an error from me. I should have started caching such witnesses on a separate spawned thread.